### DISCUSSION TITLE
Does libunwind support VDSOs

### DISCUSSION THREAD

**Discussion by araychaudhury** [libunwind/libunwind#696] [Q&A]
Hi,
 I am using libunwind-ptrace for remote process backtracing. However it fails to walk the stack when a stack frame is in VDSO . On further investigation I found tdep_get_elf_image implantation in os-linux.c tries to open VDSO as a file. As the file does not exist it fails to  step further. 

A patch to mmap the content of VDSO and then parse the content for stack walking exists in https://sourceware.org/pipermail/frysk-cvs/2008q1/007245.html . 

Any idea why this type of patch is not applied to libunwind ?

Thanks in Advance,
-Abhijit

**Comment by bregma** [comment]
Because [#658](https://github.com/libunwind/libunwind/pull/658) hasn't been merged yet?

**Reply by araychaudhury** [reply]
Hi Stephen,

I have looked into the patch:

if (strcmp(mi.path, VDSO_NAME) == 0)
    {
      /* N.B. Assumes vDSO is described in one entry in maps */
      ei->image = (void*) *segbase;
      ei->size = (void*) hi - ei->image;
      ei->mapped = 0;

      rc = elf_w (valid_object) (ei) ? 0 : -1;
    }

It will not work for me as I am remotely doing stack walk using ptrace , so "rc = elf_w (valid_object) (ei) ? 0 : -1;" will simply fail as it is not the processes own address space. It should have been memcpy'ed into the process's own address space through accessor functions and then checked for validity.

So I found  https://sourceware.org/pipermail/frysk-cvs/2008q1/007245.html patch more appealing .

Please let me know your comments .

Thanking You,
-Abhijit

**Reply by bregma** [reply]
Well, there's a reason why #658 hasn't been merged, and it's because I don't think it will work in the remotes, as indicated in the comment I left on the PR. Need a good test case to demonstrate it working (or not).

**Reply by araychaudhury** [reply]
Hi Steven,

I have tried to use the patch in my code  and it does not work for remote. 

So I am trying to pull https://sourceware.org/pipermail/frysk-cvs/2008q1/007245.html , but passing unw_addr_space_t as, and void *arg to tdep_get_elf_image from the top is proving to be too much work, especially now that it supports so many platforms .

Thanking You,
-Abhijit

**Reply by araychaudhury** [reply]
Hi Steven,

I have ported  https://sourceware.org/pipermail/frysk-cvs/2008q1/007245.html for current libunwind . I am using a 64bit x86_64 kernel and a 32 bit x86 kernel with "while(1) sleep (1)" loop.

With the changes I am able to find proper backtrace of the sleep application, which was not possible with the current libunwind :

0xf7f7c548: (__kernel_vsyscall+0x00000009)
0xf7e2248a: [ /usr/lib/i386-linux-gnu/libc-2.31.so + 0xa948a ] (__GI___clock_nanosleep_time64+0x0000007b)
0xf7e22802: [ /usr/lib/i386-linux-gnu/libc-2.31.so + 0xa9802 ] (__clock_nanosleep_2+0x00000052)
0xf7e29284: [ /usr/lib/i386-linux-gnu/libc-2.31.so + 0xb0284 ] (__GI___nanosleep+0x00000024)
0xf7e29194: [ /usr/lib/i386-linux-gnu/libc-2.31.so + 0xb0194 ] (__sleep+0x00000075)
0x565dc226: [ /home/sleep/sleep + 0x226 ] (main+0x0000003a)
0xf7d7aed4: [ /usr/lib/i386-linux-gnu/libc-2.31.so + 0x1ed4 ] (__libc_start_main+0x000000f5)
0x565dc0e4: [ /home/sleep/sleep + 0xe4 ] (_start+0x00000035)

Do you think I can submit the patch ?
[vdso.patch](https://github.com/libunwind/libunwind/files/13960982/vdso.patch)

**Reply by bregma** [reply]
Just submit a pull request so it will run through the CI pre-commit checks and I can do offline QA builds on other targets.

Make sure to give proper attribution to its original inspiration.

It would be really great if you can provide tests cases to exercise this code for both local and remote unwinding as a part of the PR.

**Reply by araychaudhury** [reply]
Hi Steven ,

Do you have any handbook on PR raising procedure ? I will also try to make armv7 , x86 and x86_64 build and check the behavior .

As I have changed function prototype across many platform , do you know of any sandbox where I can at-least build for some of the platforms .

Thanks in Advance,
-Abhijit

**Reply by bregma** [reply]
There are no documented contribution guidelines for libunwind at this time. On new sources I use `astyle -A7 -s2` but follow the style of existing source files as much as possible. Also, be very very careful of external API or ABI breaks.

I have no suggestion on any test builders. I wish I did, it would make my life easier. Any PR you submit will trigger the pre-commit CI runs, which cross-build a number of architectures on GNU/Linux and a few on Musl/Linux and but everything else is guesswork. I will run test builds against QNX 8.0 x86_64 and aarch64 as a part of any review I do.

**Reply by araychaudhury** [reply]
Hi Steven,

I am new to github PR procedure . I have submitted https://github.com/libunwind/libunwind/pull/728.

Please run Ci/CD and let me know if any changes are required .

Thanking You,
-Abhijit
